packages/core/src/StateNode.ts

```diff
@@ -220,8 +206,28 @@ export class StateNode<
     this.history =
       this.config.history === true ? 'shallow' : this.config.history || false;
 
-    this.entry = toArray(this.config.entry).slice();
-    this.exit = toArray(this.config.exit).slice();
+    const convertAction = (
+      action: UnknownAction,
+      kind: 'entry' | 'exit',
+      i: number
+    ): UnknownActionObject => {
+      if (typeof action === 'string') {
+        return { type: action };
+      }
+      if (typeof action === 'function' && !('resolve' in action)) {
+        const type = `${this.id}|${kind}:${i}`;
+        this.machine.implementations.actions[type] = action as any;
```

Comment:
1. I don't particularly like that we are writing to this object because it might be shared among different machines in some way and we might potentially leak things between instances
2. i'd be wary of the code [here](https://github.com/statelyai/xstate/blob/6341cfbb00b3767dd075ffe49466dd2e58d4f948/packages/xstate-react/src/useActorRef.ts#L36-L40), I wonder if this gets "fixed" somehow but I think there might run into stale closure hazard by resolving this here and caching it